python: use type hints

axel-h opened the pull request:
This change will require at least Python 3.7, but we have that in the official docker containers now.

lsf37 reviewed (commented):
I would be fine with this, the more type checking the better.

What python version does Ubuntu 18 default to? It's still the default build platform described on the doc site, although we should update that to 20.04LTS.

axel-h commented:
> What python version does Ubuntu 18 default to

It's 3.6.5, see https://packages.ubuntu.com/bionic/python/python3
20.04 uses 3.8.2, see https://packages.ubuntu.com/focal/python/python3

The reason I have this patch pending for a while is, that 3.7 was not available. Just with the recent change of the docker containers it is now to I thought it time to make a PR.

axel-h commented:
Rebased and extended the comments in the scripts to clarify the python version dependencies.

axel-h commented:
Just to clarify, is the merge blocker here still, that we want to keep Ubuntu 18 support where Python is at 3.6, even if we are no longer using this in CI and the docker containers?

lsf37 commented:
Yes, we still want to support Ubuntu 18 for a while (I forgot who long, I think it's only another few months)

lsf37 commented:
End of support for Ubuntu 18 is Apr 2023, so I'd leave this unmerged until then.

axel-h commented:
> End of support for Ubuntu 18 is Apr 2023, so I'd leave this unmerged until then.

It's just a few days now... ;)

axel-h committed:
python: fix return type and annotation

Also raise exception is there are no regions at all.

Signed-off-by: Axel Heider <axelheider@gmx.de>

axel-h commented:
> > End of support for Ubuntu 18 is Apr 2023, so I'd leave this unmerged until then.
> It's just a few days now... ;)

So, given https://github.com/seL4/docs/pull/191 got merged and removed Ubuntu 18.04 references, is it a good point to consider merging this PR then?

Indanz reviewed (dismissed):
Looks okay to me, however, I'm not a Python expert.

Have you tested all these changes?

Could you explain the following comment in the commit message:

> "annotations" exists in __future__ since 3.7.0b1 and was planned
> to become mandatory in 3.10. But this has been postponed.

Is still relevant?

lsf37 commented:
> Is still relevant?

It's relevant in the sense that this will fail for python 3.6 and earlier. We should check what the default versions are for the current Debian stable and Ubuntu 20. If that's >= 3.7 I think we can merge this.

axel-h commented:
Debian:
- buster (oldoldstable): 3.7.3
- bullseye (oldstable); 3.9.2
- bookworm (stable): 3.11.2
Ubuntu LTS: 
- 18.4: 3.6.5
- 20.4: 3.8.2
- 22.4: 3.10

axel-h commented:
Currently type annotations are still not checked, but there are packages that enable this. I will add a commit to this PR that enforces the check. But this may need a new docker container then also.
Making this PR a draft again for now

lsf37 reviewed (dismissed):
Happy for this to be merged now that all supported host deps have python > 3.7

axel-h commented on tools/hardware/config.py:
```diff
@@ -3,7 +3,8 @@
 #
-from typing import List, Set
+
+from __future__ import annotations
```
We need this, so we can write `def foo() -> SomeType: ...` instead of having to use quotes `def foo() -> 'SomeType': ...`. This was supposed to become possible by default in 3.10, but has been postponed.

axel-h replied:
We have Python 3.8 in the container now, so after some more testing it turn out this is no longer necessary in general. There are just some corner cases where I' trying to figure out where this is needed. Having this does not harm, but it allows using some feature newer python version will make the default. So while this is nice, there is also a certain danger to break the oldest version we support . Ideally, we'd have a `trustworthysystems/sel4:oldest` container that is used in a CI jobs also  to test thing are still fine in the oldest configuration we support.
Unless we don't have this, it might actually be better not to have this line anywhere unless it is absolutely required.

lsf37 commented:
@axel-h this is still on draft, but seems otherwise ready to merge. Are we good to go ahead on this one? If yes, can you please switch off draft mode?

axel-h commented:
The open point here is running this though a checker regularly. Currently, as @Indanz remarked, it is more or less an informative comment only and Python lacks a nice built-in was to really use this. I'm putting this out of draft state now, so it can be merged if you agree it is still a good step forward. I hope find more time to add a CI step that does a static check at least.

lsf37 commented:
> The open point here is running this though a checker regularly. Currently, as @Indanz remarked, it is more or less an informative comment only and Python lacks a nice built-in way to really use this. I'm putting this out of draft state now, so it can be merged if you agree it is still a good step forward. I hope find more time soon to add a CI step that does a static check at least.

Types as informative comments are a good step, I think we should merge this and do the checking as a second step (which also would definitely be good).

axel-h committed:
python: fix string format parameters

Signed-off-by: Axel Heider <axelheider@gmx.de>

axel-h commented:
Inspired by https://github.com/seL4/microkit/pull/101 I have added a MyPy CI step for this repo also, see https://github.com/axel-h/seL4/pull/154. Running this, it shows there are more things that are worth cleaning up. Also, some changes I made in this PR turn out to be no optimal. So I thing it's worth adding a few more changes here.

axel-h committed:
python: check object type first

Support comparisson operations for derived objects only, and fail for
anything else. Currently there is no use case that requires duck
typing and stricter type checking allows catching potential errors.

Signed-off-by: Axel Heider <axelheider@gmx.de>

axel-h committed:
python: use proper Tuple type annotation

Signed-off-by: Axel Heider <axelheider@gmx.de>

axel-h committed:
python: use Callable instead of Any

'Callable' was added in Python 3.6 in 'collections.abc'.

Signed-off-by: Axel Heider <axelheider@gmx.de>

axel-h committed:
python: pass jinja2 loader instance

Signed-off-by: Axel Heider <axelheider@gmx.de>

axel-h committed:
python: don't add None elements

Signed-off-by: Axel Heider <axelheider@gmx.de>

axel-h committed:
python: check for None

Signed-off-by: Axel Heider <axelheider@gmx.de>

axel-h committed:
python: type annotation for operations

Even if NotImplemented is of the type NotImplementedType technically,
it is not supposed to be mentioned explicitly in the return types.

Signed-off-by: Axel Heider <axelheider@gmx.de>

axel-h committed:
python: drop explicit index field

Lists are ordered and enumerate() provides the index.

Signed-off-by: Axel Heider <axelheider@gmx.de>

axel-h committed:
python: fail if path in chosen node does not exist

The DTB appears broken if an item in the chosen node can't be
resolved. Raise an error instead of silently ignoring this,
as it might be a typo that must be fixed. Otherwise, non-existing
items should be removed from the chosen node in a overlay file.

Signed-off-by: Axel Heider <axelheider@gmx.de>

lsf37 commented:
Thanks for testing it, but I don't want to mix in a checker into this PR. Adding a checker would be a TSC decision (which checker(s), what rules, which repos, what update policy etc). Happy to put it on the agenda for the next meeting, I think it would be a good idea to use one, but it requires some thinking. 

This PR is for type annotations as documentation, which is fine on its own.

So either we should merge this or close it if you don't think it makes sense without having a checker at the same time.

axel-h commented on tools/hardware/config.py:
```diff
@@ -35,14 +36,15 @@ def get_device_page_bits(self) -> int:
         return self.get_page_bits()
 
-    def align_memory(self, regions: Set[Region]) -> List[Region]:
+    def align_memory(self, regions: Set[Region]) -> (List[Region], Set[Region], int):
```
It seem the proper way to annotate do this is `Tuple[List[Region], Set[Region], int]`. So we should use this to avoid building legacy.

axel-h commented on tools/hardware/memory.py:
```diff
@@ -12,56 +16,56 @@
         return self.base > other.base
 
-    def __hash__(self):
+    def __hash__(self) -> bool:
```
must be `-> int`

axel-h commented:
> [...] I don't want to mix in a checker into this PR. Adding a checker would be a TSC decision (which checker(s), what rules, which repos, what update policy etc). [...] This PR is for type annotations as documentation, which is fine on its own.

Adding a type checker is not part of the PR. However, I think changes in this PR only make sense if they are also aligned with Python standard. It seems MyPy this is currently best tool to verify we comply with these standard. Thus I'd really prefer to update this PR to comply with MyPy at least.

axel-h committed:
python: remove unused clone() function

Removing the function because it is not use anywhere. Not cloning the
owner seems an inconsistent behavior anyway.

Signed-off-by: Axel Heider <axelheider@gmx.de>

axel-h committed:
python: use string routines

Element 0 may not exist if the string is empty.

Signed-off-by: Axel Heider <axelheider@gmx.de>

axel-h committed:
python: proper type annotation for optional None

Signed-off-by: Axel Heider <axelheider@gmx.de>

axel-h committed:
python: add type annotations

Signed-off-by: Axel Heider <axelheider@gmx.de>

axel-h committed:
python: drop return value of visit()

The return values is not used by any known callers of this function.
There is no sane generic way to handle return values on recursive
invocations, thus any visitor must handle this internally. This is
doable in Python easily, so there is no need for an additional
explicit context parameter either.

Signed-off-by: Axel Heider <axelheider@gmx.de>

lsf37 commented:
This started out as adding type annotations, but has now also has a whole bunch of other changes and is failing tests.

It's still on draft, so not clear if this should be reviewed or not and the approving reviews were for older version (I have removed these approvals now). I'm fine with adding annotations for documentation without changing behaviour. I'm not fine with other random changes in the same PR.

Adding a Python type checker and linter to CI (and which one) is a TSC decision and should probably be applied more widely and take into account the rest of the repositories. I can add it to the agenda of the next TSC meeting, but in any case it should be separate to adding annotations and fixing lints if the TSC decides that we want such lints to be fixed.

So it's not clear that to do with this PR now. If you are happy to revert to just adding annotations, I think this could be merged. Otherwise this should be closed.
